server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -46,42 +68,112 @@ public void beforeRefresh() throws IOException {
 
     /**
      * Upload new segment files created as part of the last refresh to the remote segment store.
-     * The method also deletes segment files from remote store which are not part of local filesystem.
+     * This method also uploads remote_segments_metadata file which contains metadata of each segment file uploaded.
      * @param didRefresh true if the refresh opened a new reference
-     * @throws IOException in case of I/O error in reading list of local files
      */
     @Override
-    public void afterRefresh(boolean didRefresh) throws IOException {
-        if (didRefresh) {
-            Set<String> localFiles = Set.of(storeDirectory.listAll());
-            localFiles.stream().filter(file -> !filesUploadedToRemoteStore.contains(file)).forEach(file -> {
-                try {
-                    remoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
-                    filesUploadedToRemoteStore.add(file);
-                } catch (NoSuchFileException e) {
-                    logger.info(
-                        () -> new ParameterizedMessage("The file {} does not exist anymore. It can happen in case of temp files", file),
-                        e
-                    );
-                } catch (IOException e) {
-                    // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
-                    logger.warn(() -> new ParameterizedMessage("Exception while uploading file {} to the remote segment store", file), e);
-                }
-            });
+    public void afterRefresh(boolean didRefresh) {
+        synchronized (this) {
+            try {
+                if (indexShard.shardRouting.primary()) {
+                    if (this.primaryTerm != indexShard.getOperationPrimaryTerm()) {
+                        this.primaryTerm = indexShard.getOperationPrimaryTerm();
+                        this.remoteDirectory.init();
+                    }
+                    try {
+                        String lastCommittedLocalSegmentFileName = SegmentInfos.getLastCommitSegmentsFileName(storeDirectory);
+                        if (!remoteDirectory.containsFile(
+                            lastCommittedLocalSegmentFileName,
+                            getChecksumOfLocalFile(lastCommittedLocalSegmentFileName)
+                        )) {
+                            deleteStaleCommits();
```

Comment:
Would clean up stall refresh ? Can we do this in the background?

Reply:
Clean up does not happen at each refresh. It would only be called when refresh is called during flush (as new segments_N file gets created). We can definitely preform stale commit deletion in the background. I have created a tracking issue here: https://github.com/opensearch-project/OpenSearch/issues/4315